In [223]:
import pandas as pd

In [224]:
data = pd.read_csv('dataset.csv', encoding='ISO-8859-1')

In [225]:
data.head()

identifier source_code  \
0  us-nara-amending-america-000001           A   
1  us-nara-amending-america-000002           A   
2  us-nara-amending-america-000003           A   
3  us-nara-amending-america-000004           A   
4  us-nara-amending-america-000005           A   

                                     source_citation source_index_number  \
0  Ames, Herman V. The Proposed Amendments to the...                   1   
1  Ames, Herman V. The Proposed Amendments to the...                   2   
2  Ames, Herman V. The Proposed Amendments to the...                   3   
3  Ames, Herman V. The Proposed Amendments to the...                   4   
4  Ames, Herman V. The Proposed Amendments to the...                   5   

                    title_or_description_from_source date_approximation  \
0                 Reservation of nondelegated powers              circa   
1                   Apportionment of Representatives              circa   
2  Restriction on Federal control over election o...              circa   
3       Restriction upon the levying of direct taxes              circa   
4                   Commercial monopolies prohibited              circa   

     year  month  day  congress congressional_session  \
0  1788.0    NaN  NaN       NaN                   NaN   
1  1788.0    NaN  NaN       NaN                   NaN   
2  1788.0    NaN  NaN       NaN                   NaN   
3  1788.0    NaN  NaN       NaN                   NaN   
4  1788.0    NaN  NaN       NaN                   NaN   

  joint_resolution_chamber  joint_resolution_number sponsor_name  \
0                      NaN                      NaN          NaN   
1                      NaN                      NaN          NaN   
2                      NaN                      NaN          NaN   
3                      NaN                      NaN          NaN   
4                      NaN                      NaN          NaN   

  sponsor_state_or_territory committee_of_referral  \
0                        NaN                   NaN   
1                        NaN                   NaN   
2                        NaN                   NaN   
3                        NaN                   NaN   
4                        NaN                   NaN   

                   last_modified  
0  2016-02-25T17:29:59.732-05:00  
1  2016-02-25T17:29:59.732-05:00  
2  2016-02-25T17:29:59.732-05:00  
3  2016-02-25T17:29:59.732-05:00  
4  2016-02-25T17:29:59.732-05:00

In [226]:
# Filter out duplicates in the 'title_or_description_from_source' column, keep the latest entry
data_unique = data.copy().drop_duplicates(subset='title_or_description_from_source', keep='last')
print(f'Dropped {data.shape[0] - data_unique.shape[0]} duplicate entries based on title_or_description_from_source.')

Dropped 7066 duplicate entries based on title_or_description_from_source.


In [227]:
data_unique.head()

identifier source_code  \
7   us-nara-amending-america-000008           A   
20  us-nara-amending-america-000021           A   
23  us-nara-amending-america-000024           A   
29  us-nara-amending-america-000030           A   
34  us-nara-amending-america-000035           A   

                                      source_citation source_index_number  \
7   Ames, Herman V. The Proposed Amendments to the...                   8   
20  Ames, Herman V. The Proposed Amendments to the...                  21   
23  Ames, Herman V. The Proposed Amendments to the...                  24   
29  Ames, Herman V. The Proposed Amendments to the...                  30   
34  Ames, Herman V. The Proposed Amendments to the...                  35   

            title_or_description_from_source date_approximation    year  \
7              Trial by jury in civil action              circa  1788.0   
20              Trial by jury in civil cases              circa  1788.0   
23                                  Religion              circa  1788.0   
29          Publication of journals annually              circa  1788.0   
34  Period of enlistment of soldiers limited              circa  1788.0   

    month  day  congress congressional_session joint_resolution_chamber  \
7     NaN  NaN       NaN                   NaN                      NaN   
20    NaN  NaN       NaN                   NaN                      NaN   
23    NaN  NaN       NaN                   NaN                      NaN   
29    NaN  NaN       NaN                   NaN                      NaN   
34    NaN  NaN       NaN                   NaN                      NaN   

    joint_resolution_number sponsor_name sponsor_state_or_territory  \
7                       NaN          NaN                        NaN   
20                      NaN          NaN                        NaN   
23                      NaN          NaN                        NaN   
29                      NaN          NaN                        NaN   
34                      NaN          NaN                        NaN   

   committee_of_referral                  last_modified  
7                    NaN  2016-02-25T17:29:59.732-05:00  
20                   NaN  2016-02-25T17:29:59.732-05:00  
23                   NaN  2016-02-25T17:29:59.732-05:00  
29                   NaN  2016-02-25T17:29:59.732-05:00  
34                   NaN  2016-02-25T17:29:59.732-05:00

In [228]:
data_unique['date_approximation'].unique()

array(['circa', nan], dtype=object)

In [229]:
# drop unnecessary columns
dropped = data_unique.drop(['identifier', 'source_code', 'date_approximation'],axis = 1)
dropped.head()

source_citation source_index_number  \
7   Ames, Herman V. The Proposed Amendments to the...                   8   
20  Ames, Herman V. The Proposed Amendments to the...                  21   
23  Ames, Herman V. The Proposed Amendments to the...                  24   
29  Ames, Herman V. The Proposed Amendments to the...                  30   
34  Ames, Herman V. The Proposed Amendments to the...                  35   

            title_or_description_from_source    year  month  day  congress  \
7              Trial by jury in civil action  1788.0    NaN  NaN       NaN   
20              Trial by jury in civil cases  1788.0    NaN  NaN       NaN   
23                                  Religion  1788.0    NaN  NaN       NaN   
29          Publication of journals annually  1788.0    NaN  NaN       NaN   
34  Period of enlistment of soldiers limited  1788.0    NaN  NaN       NaN   

   congressional_session joint_resolution_chamber  joint_resolution_number  \
7                    NaN                      NaN                      NaN   
20                   NaN                      NaN                      NaN   
23                   NaN                      NaN                      NaN   
29                   NaN                      NaN                      NaN   
34                   NaN                      NaN                      NaN   

   sponsor_name sponsor_state_or_territory committee_of_referral  \
7           NaN                        NaN                   NaN   
20          NaN                        NaN                   NaN   
23          NaN                        NaN                   NaN   
29          NaN                        NaN                   NaN   
34          NaN                        NaN                   NaN   

                    last_modified  
7   2016-02-25T17:29:59.732-05:00  
20  2016-02-25T17:29:59.732-05:00  
23  2016-02-25T17:29:59.732-05:00  
29  2016-02-25T17:29:59.732-05:00  
34  2016-02-25T17:29:59.732-05:00

<h1> Prediction: Number of Amendment Proposal w/ factors like Decades, Politcal Party in Power at the time, and whether the country was in war

In [230]:
# convert year to decades
dropped['decades'] = (dropped['year'] // 10) * 10
dropped.head()

source_citation source_index_number  \
7   Ames, Herman V. The Proposed Amendments to the...                   8   
20  Ames, Herman V. The Proposed Amendments to the...                  21   
23  Ames, Herman V. The Proposed Amendments to the...                  24   
29  Ames, Herman V. The Proposed Amendments to the...                  30   
34  Ames, Herman V. The Proposed Amendments to the...                  35   

            title_or_description_from_source    year  month  day  congress  \
7              Trial by jury in civil action  1788.0    NaN  NaN       NaN   
20              Trial by jury in civil cases  1788.0    NaN  NaN       NaN   
23                                  Religion  1788.0    NaN  NaN       NaN   
29          Publication of journals annually  1788.0    NaN  NaN       NaN   
34  Period of enlistment of soldiers limited  1788.0    NaN  NaN       NaN   

   congressional_session joint_resolution_chamber  joint_resolution_number  \
7                    NaN                      NaN                      NaN   
20                   NaN                      NaN                      NaN   
23                   NaN                      NaN                      NaN   
29                   NaN                      NaN                      NaN   
34                   NaN                      NaN                      NaN   

   sponsor_name sponsor_state_or_territory committee_of_referral  \
7           NaN                        NaN                   NaN   
20          NaN                        NaN                   NaN   
23          NaN                        NaN                   NaN   
29          NaN                        NaN                   NaN   
34          NaN                        NaN                   NaN   

                    last_modified  decades  
7   2016-02-25T17:29:59.732-05:00   1780.0  
20  2016-02-25T17:29:59.732-05:00   1780.0  
23  2016-02-25T17:29:59.732-05:00   1780.0  
29  2016-02-25T17:29:59.732-05:00   1780.0  
34  2016-02-25T17:29:59.732-05:00   1780.0

In [231]:
# make decades a column in the df
counts_df = dropped.groupby('decades').count()
counts_df = counts_df.reset_index()
counts_df

decades  source_citation  source_index_number  \
0    1780.0              175                  176   
1    1790.0               20                   20   
2    1800.0               37                   37   
3    1810.0               41                   41   
4    1820.0               77                   78   
5    1830.0               80                   81   
6    1840.0               45                   45   
7    1850.0               15                   15   
8    1860.0              412                  412   
9    1870.0              138                  138   
10   1880.0              179                  179   
11   1890.0              108                  108   
12   1900.0              143                  143   
13   1910.0              295                  295   
14   1920.0              292                  292   
15   1930.0              428                  428   
16   1940.0              226                  226   
17   1950.0              364                  364   
18   1960.0              357                  370   
19   1970.0              469                  470   
20   1980.0              294                  294   
21   1990.0              280                   10   
22   2000.0              124                    0   
23   2010.0              115                    0   

    title_or_description_from_source  year  month  day  congress  \
0                                176   176     85   85         3   
1                                 20    20     14   14        11   
2                                 37    37     27   27        26   
3                                 41    41     19   19        18   
4                                 78    78     63   63        56   
5                                 81    81     39   39        39   
6                                 45    45     40   40        39   
7                                 15    15     12   12        12   
8                                412   412    197  197       192   
9                                138   138    122  122       122   
10                               179   179    164  164       164   
11                               108   108    108  108       108   
12                               143   143    143  143       143   
13                               295   295    295  295       295   
14                               292   292    292  292       292   
15                               428   428    428  428       428   
16                               226   226    226  226       226   
17                               364   364    364  364       364   
18                               370   370    370  370       370   
19                               470   470    468  468       470   
20                               293   294    294  294       294   
21                               280   280    280  280       280   
22                               124   124    124  124       124   
23                               115   115    115  115       115   

    congressional_session  joint_resolution_chamber  joint_resolution_number  \
0                       3                         0                        0   
1                      11                         0                        0   
2                      26                         0                        0   
3                      18                         0                        0   
4                      56                         0                        0   
5                      39                         0                        0   
6                      39                         0                        0   
7                      12                         0                        0   
8                     192                         0                        0   
9                     122                         0                        0   
10                    164                        14                       14   
11               

In [232]:
# create new df with only necessary columns we will be using
decades_df = pd.DataFrame({
    'decades' : counts_df['decades'],
    'proposal_counts' : counts_df['title_or_description_from_source']
    
})
decades_df

decades  proposal_counts
0    1780.0              176
1    1790.0               20
2    1800.0               37
3    1810.0               41
4    1820.0               78
5    1830.0               81
6    1840.0               45
7    1850.0               15
8    1860.0              412
9    1870.0              138
10   1880.0              179
11   1890.0              108
12   1900.0              143
13   1910.0              295
14   1920.0              292
15   1930.0              428
16   1940.0              226
17   1950.0              364
18   1960.0              370
19   1970.0              470
20   1980.0              293
21   1990.0              280
22   2000.0              124
23   2010.0              115

In [233]:
# create new dataset for external information on war & political power
political_power = pd.read_excel('pol_power.xlsx')
political_power['decades'] = political_power['decades'].astype(float)
political_power.head()

decades  in war?  \
0   1780.0        1   
1   1790.0        1   
2   1800.0        1   
3   1810.0        1   
4   1820.0        0   

  politcal party in power (democratic, republican, neither, or both)  
0                                            neither                  
1                                            neither                  
2                                               both                  
3                                               both                  
4                                               both

In [234]:
# add new info into our existing df
decades_df['in_war'] = political_power['in war?']
decades_df['politcal_power'] = political_power['politcal party in power (democratic, republican, neither, or both)']
decades_df

decades  proposal_counts  in_war politcal_power
0    1780.0              176       1        neither
1    1790.0               20       1        neither
2    1800.0               37       1           both
3    1810.0               41       1           both
4    1820.0               78       0           both
5    1830.0               81       1     democratic
6    1840.0               45       1     democratic
7    1850.0               15       1     democratic
8    1860.0              412       1     republican
9    1870.0              138       1     republican
10   1880.0              179       1     republican
11   1890.0              108       1     republican
12   1900.0              143       0     republican
13   1910.0              295       1     democratic
14   1920.0              292       0     republican
15   1930.0              428       0     democratic
16   1940.0              226       1     democratic
17   1950.0              364       1     republican
18   1960.0              370       1     democratic
19   1970.0              470       1           both
20   1980.0              293       0     republican
21   1990.0              280       1     democratic
22   2000.0              124       1     republican
23   2010.0              115       1     democratic

In [235]:
# one hot encode politcal party for modeling
df_encoded = pd.get_dummies(decades_df, columns=['politcal_power'])
df_encoded = df_encoded.drop(['politcal_power_both'], axis = 1) # to avoid multicollinearity
df_encoded

decades  proposal_counts  in_war  politcal_power_democratic  \
0    1780.0              176       1                      False   
1    1790.0               20       1                      False   
2    1800.0               37       1                      False   
3    1810.0               41       1                      False   
4    1820.0               78       0                      False   
5    1830.0               81       1                       True   
6    1840.0               45       1                       True   
7    1850.0               15       1                       True   
8    1860.0              412       1                      False   
9    1870.0              138       1                      False   
10   1880.0              179       1                      False   
11   1890.0              108       1                      False   
12   1900.0              143       0                      False   
13   1910.0              295       1                       True   
14   1920.0              292       0                      False   
15   1930.0              428       0                       True   
16   1940.0              226       1                       True   
17   1950.0              364       1                      False   
18   1960.0              370       1                       True   
19   1970.0              470       1                      False   
20   1980.0              293       0                      False   
21   1990.0              280       1                       True   
22   2000.0              124       1                      False   
23   2010.0              115       1                       True   

    politcal_power_neither  politcal_power_republican  
0                     True                      False  
1                     True                      False  
2                    False                      False  
3                    False                      False  
4                    False                      False  
5                    False                      False  
6                    False                      False  
7                    False                      False  
8                    False                       True  
9                    False                       True  
10                   False                       True  
11                   False                       True  
12                   False                       True  
13                   False                      False  
14                   False                       True  
15                   False                      False  
16                   False                      False  
17                   False                       True  
18                   False                      False  
19                   False                      False  
20                   False                       True  
21                   False                      False  
22                   False                       True  
23                   False                      False

<h2> Modeling: Linear Regression

In [236]:
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error, r2_score

In [237]:
X = df_encoded.drop(columns=['proposal_counts', 'decades'])  # predictors
y = df_encoded['proposal_counts'] # target

In [238]:
model = LinearRegression()
model.fit(X, y)

# View coefficients
coefficients = pd.Series(model.coef_, index=X.columns)
print("Intercept:", model.intercept_)
print("Coefficients:")
print(coefficients)

Intercept: 195.0534351145039
Coefficients:
in_war                      -51.404580
politcal_power_democratic    56.750636
politcal_power_neither      -45.648855
politcal_power_republican    67.327396
dtype: float64


In [239]:
y_pred = model.predict(X)
mse = mean_squared_error(y, y_pred)
r2 = r2_score(y, y_pred)

print("Mean Squared Error:", mse)
print("R-squared:", r2)

Mean Squared Error: 17271.552728300823
R-squared: 0.0983930194254975


R-squared of 0.098 means the model only explains about 9.8% of the variance in proposal_counts, suggesting weak explanatory power in our variables. This means political power and war does not explain the number of amendment proposal count that well. 

In [240]:
import numpy as np
avg_prop_count = df_encoded['proposal_counts'].mean()
RMSE = np.sqrt(mse)

print(avg_prop_count, RMSE)

197.08333333333334 131.4212795870624


In [241]:
relative_error = RMSE/avg_prop_count
print(relative_error)

0.6668310169322405


On average, the model’s predictions are off by about 131 proposals per decade.
The relative eroor suggests that model’s average error is about two-thirds the size of the actual average value, which is pretty high, again suggesting unreliability of our model.

<h2> Modeling: Poisson Regression

In [242]:
import statsmodels.api as sm
import statsmodels.formula.api as smf

model = smf.glm(formula='proposal_counts ~ in_war + politcal_power_democratic + politcal_power_neither + politcal_power_republican',
                data=df_encoded, family=sm.families.Poisson()).fit()
print(model.summary())

                 Generalized Linear Model Regression Results                  
Dep. Variable:        proposal_counts   No. Observations:                   24
Model:                            GLM   Df Residuals:                       19
Model Family:                 Poisson   Df Model:                            4
Link Function:                    Log   Scale:                          1.0000
Method:                          IRLS   Log-Likelihood:                -1198.7
Date:                Wed, 25 Jun 2025   Deviance:                       2235.1
Time:                        18:57:08   Pearson chi2:                 2.28e+03
No. Iterations:                     5   Pseudo R-squ. (CS):              1.000
Covariance Type:            nonrobust                                         
                                        coef    std err          z      P>|z|      [0.025      0.975]
-----------------------------------------------------------------------------------------------------
Interc

In [243]:
democrat = np.exp(0.3104)
neither = np.exp(-0.4041)
republican = np.exp(0.3563)
war = np.exp(-0.2346)

print(democrat, neither, republican, war)

1.3639705932663844 0.6675773601949113 1.4280358947770884 0.7908871412226887


Given the Poisson Regression summary, we can conclude the following:
1. proposal count is 36% higher in democratic eras compared to when both parties are in power
2. proposal count is 33% lower when neither parties are in power
3. proposal counts is 43% higher in republican eras compared to when both parties are in power
4. during war periods, proposal counts are 21% lower on average

all of the coefficients have very low p-values (P>|z| = 0.000), meaning they’re statistically significant at any standard threshold (0.05, 0.01, etc). this means that we can rely on the interpretations above to make a prediction on amendment proposal counts based on politcal party in power and war status of the country at the time.